### PR TITLE
Increase zfs_vdev_async_write_min_active to 2

### PR DESCRIPTION
**Pull request by DeHackEd**
### Description
Resilver operations on frequently cause only a small amount of dirty data
to be writte to disk at a time, resulting in the IO scheduler to only
use 1 write at a time to the resilvering disk. When it is rotational media
the drive will often travel past the next sector to be written before
receiving a write command from ZFS, significantly delaying the write of
the next sector.

Raise zfs_vdev_async_write_min_active so that drives are kept fed
during resilvering.

### Motivation and Context
This has been brought up in other issues, such as #4825, and on IRC a few times. Resilvers are slow due to what appears to be a drive only receiving 1 write operation at a time which kills sequential writes during resilvers.

### How Has This Been Tested?
Several instances of people trying it have resulted in immediate throughput increases for Resilvers. No other testing has been done. See the above mentioned issue for a specific instance including the performance improvement.

### Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist:
- [X] My code follows the ZFS on Linux code style requirements.
- [X] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by DeHackEd** [comment]
I have opened this issue half to spur discussion on the topic and half as a serious consideration to have the limit raised.

**Comment by ahrens** [comment]
The change seems fine, but can you explain the mechanism by which this has an effect on resilver performance?  The scrub/resilver i/os are issued as ZIO_PRIORITY_SCRUB which uses zfs_vdev_scrub_max_active=2.

**Comment by RubenKelevra** [comment]
@ahrens the values 1 and 8 has been discussed here:
https://github.com/zfsonlinux/zfs/issues/4825#issuecomment-289247129

The values 2 and 3 has been discussed here: 
https://github.com/zfsonlinux/zfs/issues/4825#issuecomment-289813176

The different latencys for the values 1-8 has been plotted by me on a hypervisor with several VMs running contains a lot of database operations:
https://github.com/zfsonlinux/zfs/issues/4825#issuecomment-291656265

An overview about our findings can be found here:
https://github.com/zfsonlinux/zfs/issues/4825#issuecomment-291934140

**Comment by DeHackEd** [comment]
The scrub/resilver uses those for reads, but the self-repair write appears to be async writes. The repair of a resilver uses the same code that repairs damaged blocks, right?

Eg: https://github.com/zfsonlinux/zfs/blob/master/module/zfs/vdev_mirror.c#L590

**Comment by ahrens** [comment]
That sounds right to me.  Would you like to add a comment explaining why decreasing zfs_vdev_async_write_min_active below 2 can cause poor performance?

**Comment by DeHackEd** [comment]
Updated the man page. How does this wording sound? Is the comment about using a value of '3' too wordy?

**Comment by kpande** [comment]
not wordy to me, I like it. succinct and informative.

**Comment by behlendorf** [comment]
One possible way to eek out that additional performance of a default value of `3` would be to detect this case in `vdev_queue_max_async_writes()`.  The code already dynamically raises the limit when there are pending sync tasks.  A similar thing could be done when we know there are outstanding resilver IOs.

Below is a proposed patch which increases the allowed number of outstanding async writes while `dsl_scan_sync()` is actively resilvering.  I haven't had a chance to verify if there's a practical benefit to this, but it should help.  It would be great to get some data which confirms that.

This change, or something similar, can always be done at a latter date and doesn't need to be part of this PR.  But I wanted to post the patch for comment.

```diff
diff --git a/module/zfs/vdev_queue.c b/module/zfs/vdev_queue.c
index bf2e24c..d22f29d 100644
--- a/module/zfs/vdev_queue.c
+++ b/module/zfs/vdev_queue.c
@@ -33,6 +33,7 @@
 #include <sys/zio.h>
 #include <sys/avl.h>
 #include <sys/dsl_pool.h>
+#include <sys/dsl_scan.h>
 #include <sys/metaslab_impl.h>
 #include <sys/spa.h>
 #include <sys/spa_impl.h>
@@ -280,6 +281,14 @@ vdev_queue_max_async_writes(spa_t *spa)
        if (spa_has_pending_synctask(spa))
                return (zfs_vdev_async_write_max_active);
 
+       /*
+        * When actively resilvering allow the IOs to be written as quickly
+        * as possible in order to minimize rebuild time.
+        */
+       if (dsl_scan_resilvering(dp) && dsl_scan_active(dp->dp_scan) &&
+           dp->dp_scan->scn_zio_root != NULL)
+               return (zfs_vdev_async_write_max_active);
+
        dirty = dp->dp_dirty_total;
        if (dirty < min_bytes)
                return (zfs_vdev_async_write_min_active);
```

**Comment by DeHackEd** [comment]
Oh I like that idea.

My only question is whether there is any benefit to non-resilver workloads to raise the limit anyway. I honestly don't know.

**Comment by ryao** [comment]
@behlendorf That looks good to me.

**Comment by behlendorf** [comment]
As a general rule I'd like to keep the number of special cases to a minimum.  The limit does increase linearly with the amount of dirty data which should handle user generated workloads.  As for scrubbing my expectation is that repair IOs should be the exception, not the norm, so the patch limits this tweak to resilvering.

**Comment by RubenKelevra** [comment]
@DeHack

**Comment by RubenKelevra** [comment]
@DeHackEd I think there is an improvement for non-resilvering workloads, since we're feeding NCQ much better random IO is slightly better on writes, like showed in my case. Also any stream of data should be enhanced, since we don't waste time anymore writing with small gaps between each actual disk-write.

**Comment by ahrens** [comment]
@RubenKelevra Normal write workloads of should automatically cause the number of outstanding async writes to increase beyond 1 when more write performance is needed.

**Comment by RubenKelevra** [comment]
> @RubenKelevra Normal write workloads of should automatically cause the number of outstanding async writes to increase beyond 1 when more write performance is needed.
> 


Alright, I just observed a slight reduction of usage on my disks while increasing the minimum async write. But it was marginal, maybe 10%.


Best regards

Ruben
